### ISSUE TITLE
Could not load jasper fonts after upgrade to spring boot 2.0.0

### ISSUE THREAD

**Issue by maoyach** [spring-projects/spring-boot#12504]
It's OK with spring boot 1.5.9. After upgrade to 2.0.0, get an error:
net.sf.jasperreports.engine.fonts.InvalidFontException: Error loading font "fonts/宋体/simsun.ttf".

I think it's a problem of "LaunchedURLClassLoader". This ClassLoader can't load source from a package path containing Chinese characters.

Finnally, after change package from "fonts/宋体/simsun.ttf" to "fonts/simsun/simsun.ttf", the "LaunchedURLClassLoader" can load the font correctly.

Hope to fix it.

**Comment by wilkinsona** [comment]
We have tests for accessing entries with a non-ASCII name from a fat jar such as [this one](https://github.com/spring-projects/spring-boot/blob/ac644416118193cee9f1b945bf6c75e8fc266e1d/spring-boot-project/spring-boot-tools/spring-boot-loader/src/test/java/org/springframework/boot/loader/jar/JarFileTests.java#L135) so there may be more to this.

Can you please provide a small sample that reproduces the problem you have described?

**Comment by maoyach** [comment]
Sample code is like this:
		ClassLoader classLoader = Thread.currentThread().getContextClassLoader(); 
		System.out.println(classLoader);
		
		// This get null with "java -jar", but get right url when debug in eclipse 
		URL url = classLoader.getResource("fonts/宋体/simsun.ttf");
		System.out.println(url);
		
		// This can get right url both with "java -jar" and debug in eclipse
		url = classLoader.getResource("fonts/simhei/simhei.ttf");
		System.out.println(url);

The package structure is like this:
![image](https://user-images.githubusercontent.com/37408766/37472238-9313871e-28a6-11e8-9c98-5a8d14de7737.png)

**Comment by philwebb** [comment]
@maoyach Could you please share an actual project that we can clone and run.

**Comment by maoyach** [comment]
This is a simple eclipse project to show the issue:
[getresource.zip](https://github.com/spring-projects/spring-boot/files/1817665/getresource.zip)

**Comment by snicoll** [comment]
@maoyach I get `null` in IntelliJ IDEA with both Spring Boot `2.0.0.RELEASE` and `1.5.9.RELEASE` (it may be a local specific issue though because that directory shows up as `ÀŒÃÂ` for me).

**Comment by wilkinsona** [comment]
I get `null` in Eclipse too, but I think that's because the zip has mangled the `宋体` folder and it's become `ÀŒÃÂ`. If I rename the folder to `宋体`, I get the following output in Eclipse:

```
sun.misc.Launcher$AppClassLoader@4e25154f
file:/Users/awilkinson/Downloads/getresource/target/classes/fonts/%e5%ae%8b%e4%bd%93/simsun.ttf
file:/Users/awilkinson/Downloads/getresource/target/classes/fonts/simhei/simhei.ttf
```

If I then package and run the jar, I get this output:

```
org.springframework.boot.loader.LaunchedURLClassLoader@17f052a3
null
jar:file:/Users/awilkinson/Downloads/getresource/target/testgetresource.jar!/BOOT-INF/classes!/fonts/simhei/simhei.ttf
```

I think that means I've successfully reproduced the problem.

**Comment by wilkinsona** [comment]
>  I'd guess it's been caused by the performance optimisations in the launcher.

Nope. It stopped working in 2.0.0.RC1 and the performance optimisations were in 2.0.0.RELEASE

**Comment by wilkinsona** [comment]
The regression's due to the change made to `AsciiBytes.hashCode()` in [this commit](https://github.com/spring-projects/spring-boot/commit/aa66d5dfb83fb04139688df8411b15b9c3b6f29d). The following test reproduces the root cause of the problem:

```
@Test
public void hashCodeFromInstanceMatchesHashCodeFromString() {
	String name = "fonts/宋体/simsun.ttf";
	assertThat(new AsciiBytes(name).hashCode()).isEqualTo(AsciiBytes.hashCode(name));
}
```

**Comment by wilkinsona** [comment]
The new matches method introduced in [the same commit] also seems to have problems with UTF-8 characters. The following test fails at the moment:

```java
@Test
public void instanceCreatedFromCharSequenceMatchesSameCharSequence() {
	String name = "fonts/宋体/simsun.ttf";
	assertThat(new AsciiBytes(name).matches(name, NO_SUFFIX)).isTrue();
}
```
